File: graph/find_eventual_safe_states.py
from typing import List
from collections import deque


class Solution:
    # dfs
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        res = []
        safe = {}

        def dfs(node):
            if node in safe:
                return safe[node]
            safe[node] = False
            # if there is no nei then it is safe
            # or if all nei nodes are safe then it is safe
            for neiNode in graph[node]:
                if not dfs(neiNode):
                    return False
            safe[node] = True
            return True

        for node in range(n):
            if dfs(node):
                res.append(node)

        return res

    # bfs(topological sort)
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        N = len(graph)
        indegree_edges = [[] for i in range(N)]
        outdegree = [0] * N

        for i in range(N):
            outdegree[i] = len(graph[i])
            for j in graph[i]:
                indegree_edges[j].append(i)
        
        q = deque()
        res = [] * N
        for node in range(N):
            if outdegree[node] == 0:
                q.append(node)

        while q:
            cur_node = q.popleft()
            res.append(cur_node)
            for nei_node in indegree_edges[cur_node]:
                if outdegree[nei_node] != 0:
                    outdegree[nei_node] -= 1
                if outdegree[nei_node] == 0:
                    q.append(nei_node)

        return sorted(res)

File: graph/find_eventual_safe_states.java
public class find_eventual_safe_states {
    
}

class Solution {
    // dfs
    public List<Integer> eventualSafeNodes1(int[][] graph) {
        Map<Integer, Boolean> safe = new HashMap<>();
        List<Integer> res = new ArrayList<>();
        for(int i = 0; i < graph.length; i++){
            if(dfs(i, safe, graph)){
                res.add(i);
            }
        }
        return res;
    }

    private boolean dfs(int i, Map<Integer, Boolean> safe, int[][] graph){
        if(safe.containsKey(i)){
            return safe.get(i);
        }
        safe.put(i, false);
        for(int j = 0; j < graph[i].length; j++){
            int nei = graph[i][j];
            if(!dfs(nei, safe, graph)){
                return safe.get(nei);
            }
        }
        safe.put(i, true);
        return true;
    }

    // bfs
    public List<Integer> eventualSafeNodes2(int[][] graph){
        int N = graph.length;
        List<List<Integer>> edges = new ArrayList<>(N);
        for(int i = 0; i < N; i++){
            edges.add(new ArrayList());
        }
        int[] indegree = new int[N];
        for(int i = 0; i < N; i++){
            indegree[i] = graph[i].length;
            for (int j : graph[i]) {
                edges.get(j).add(i);
            }
        }

        Deque<Integer> queue = new ArrayDeque<>();
        List<Integer> res = new ArrayList<>();

        for(int i = 0; i < N; i++){
            if(indegree[i] == 0){
                queue.add(i);
            }
        }

        while(!queue.isEmpty()){
            int u = queue.poll();
            res.add(u);
            for (int i : edges.get(u)) {
                if (indegree[i] != 0) {
                    indegree[i]--;
                    if (indegree[i] == 0) {
                        queue.add(i);
                    }
                }
            }
        }

        Collections.sort(res);
        return res;
    }
}